File: Simulador/simulador.py
print("Simulador de Memória")

memoria = ['0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000']

menu = '' # declara a variável menu como vazia

while menu != 'P': # condição while de parada
  
    menu = input("Escolha uma Opção: \n E - Escrever \n L - Ler \n S - Status da memória \n P - Parar \n : ") # menu de opções

    # Parte 1: Operação de Leitura
    if menu == 'L': 
      
        # Vai pedir para o usuário escolher o endereço
        bin_endereco_leitura = input("Escolha um endereço: ")

        # Condicional 
        if len(bin_endereco_leitura) == 4: # Vai estabelecer a condição do tamanho do endereço inserido
            endereco_leitura = int(bin_endereco_leitura, 2) # Transforma o endereço inserido para o formato binário
          
            if 0 <= endereco_leitura <= 15: # Vai estabelecer a condição se o endereço está dentro do intervalo correto 
                print("Endereço", bin_endereco_leitura, "->", memoria[endereco_leitura]) # Vai retornar o endereço e o valor armazenado nele
              
            else: # Mensagem informando que o endereço é inválido
                print("Endereço inválido. Digite um endereço binário: ")
              
        else: # Mensagem informando que o tamanho do endereço está incorreto
            print("Endereço inválido. Digite um endereço binário: ")

    # Parte 2: Operação de Escrita
    elif menu == 'E':
        bin_endereco_escrita = input("Escolha um endereço: ")
      
        if len(bin_endereco_escrita) == 4: # Condição para verificar o tamanho do endereço
            endereco_escrita = 0 # inicia a variavel com valor 0 pra evitar erro

            # Transformar o endereço que foi inserido em binário para decimal indicando o índice da lista "memoria"
            endereco_escrita = int(bin_endereco_escrita, 2) # Transforma o endereço inserido para o formato binário
                  
            if 0 <= endereco_escrita <= 15:
                valor = input("Digite o valor em binário a ser armazenado: ")
              
                if len(valor) == 4:
                    memoria[endereco_escrita] = valor # Atribui o valor digitado ao índice em "memoria"
                  
                    print("Valor", valor, "armazenado no endereço", bin_endereco_escrita, ".") # Retorna o valor armazenado na memória e o local
                
                else:
                    print("Valor incorreto. Digite um valor em binário: ") # Mensagem de erro caso tenha menos de 4 caracteres
                  
            else:
                print("Endereço inválido. Digite um endereço binário: ") # Mensagem de erro caso esteja fora do intervalo
              
        else:
            print("Endereço inválido. Digite um endereço binário: ") 

    # Parte 3: Mostrar o Status da Memória
    elif menu == 'S':
        print("Status da memória: ")
        for i in range(16):
            endereco_binario = format(i, '04b') 
            print(endereco_binario, "-", memoria[i])

    # Parte 4: Condição de parada
    elif menu == 'P':
        print("Obrigada. Simulação Finalizada.")

    # Parte 5: Mensagem de opção inválida
    else:
        print("Opção inválida. Escolha uma opção válida.")

#


# variáveis do código: 

'''
valor: valor a ser armazenado

memoria: lista com índices de endereços da memória

e_endereco: endereço inserido pelo usuário em binário para operação de escrita

bin_e_endereco: endereço já transformado para binário para relacionar com o índice em "memoria" na operação de escrita

l_endereco: endereço inserido pelo usuário em binário para operação de leitura

bin_l_endereco: endereço já transformado para binário para relacionar com o índice em "memoria" na operação de leitura

menu: menu de opções
'''

File: Simulador/simulador.js
console.log("Simulador de Memória");
console.log();
let memoria = ['0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000', '0000'];
let palavra = 4;

let menu = '';

while (menu !== 'P') {
    menu = prompt("Escolha uma Opção: \n E - Escrever \n L - Ler \n S - Status da memória \n P - Parar \n : ");
    console.log();

    if (menu === 'L') {
        let bin_endereco_leitura = prompt("Escolha um endereço: ");
        console.log();

        if (bin_endereco_leitura.length === palavra) {
            let endereco_leitura = parseInt(bin_endereco_leitura, 2);

            if (0 <= endereco_leitura <= 15) {
                console.log("Endereço " + bin_endereco_leitura + " -> " + memoria[endereco_leitura]);
                console.log();
            } else {
                console.log("Endereço inválido. Digite um endereço binário: ");
            }
        } else {
            console.log("Endereço inválido. Digite um endereço binário: ");
        }

    } else if (menu === 'E') {
        let bin_endereco_escrita = prompt("Escolha um endereço: ");

        if (bin_endereco_escrita.length === palavra) {
            let endereco_escrita = parseInt(bin_endereco_escrita, 2);

            if (0 <= endereco_escrita <= 15) {
                let valor = prompt("Digite o valor em binário a ser armazenado: ");
                console.log();

                if (valor.length === palavra) {
                    memoria[endereco_escrita] = valor;
                    console.log();
                    console.log("Valor " + valor + " armazenado no endereço " + bin_endereco_escrita + ".");
                    console.log();
                } else {
                    console.log("Valor incorreto. Digite um valor em binário: ");
                }
            } else {
                console.log("Endereço inválido. Digite um endereço binário: ");
            }
        } else {
            console.log("Endereço inválido. Digite um endereço binário: ");
        }

    } else if (menu === 'S') {
        console.log("Status da memória: ");
        for (let i = 0; i < 16; i++) {
            let endereco_binario = i.toString(2).padStart(4, '0');
            console.log(endereco_binario + " - " + memoria[i]);
        }

    } else if (menu === 'P') {
        console.log("Obrigada. Simulação Finalizada.");

    } else {
        console.log("Opção inválida. Escolha uma opção válida.");
    }
}


// variáveis do código: 

/*
valor: valor a ser armazenado

memoria: lista com índices de endereços da memória

e_endereco: endereço inserido pelo usuário em binário para operação de escrita

bin_e_endereco: endereço já transformado para binário para relacionar com o índice em "memoria" na operação de escrita

l_endereco: endereço inserido pelo usuário em binário para operação de leitura

bin_l_endereco: endereço já transformado para binário para relacionar com o índice em "memoria" na operação de leitura

menu: menu de opções
*/